Mathlib/Data/Multiset/Fintype.lean

```diff
@@ -176,59 +174,136 @@ def Multiset.coeEquiv (m : Multiset α) : m ≃ m.toEnumFinset where
+theorem map_univ_coeEmbedding (m : Multiset α) :
     (Finset.univ : Finset m).map m.coeEmbedding = m.toEnumFinset := by
   ext ⟨x, i⟩
   simp only [Fin.exists_iff, Finset.mem_map, Finset.mem_univ, Multiset.coeEmbedding_apply,
     Prod.mk.inj_iff, exists_true_left, Multiset.exists_coe, Multiset.coe_mk, Fin.val_mk,
     exists_prop, exists_eq_right_right, exists_eq_right, Multiset.mem_toEnumFinset, true_and]
 
 @[simp]
-theorem Multiset.map_univ_coe (m : Multiset α) :
+theorem map_univ_coe (m : Multiset α) :
     (Finset.univ : Finset m).val.map (fun x : m ↦ (x : α)) = m := by
   have := m.map_toEnumFinset_fst
   rw [← m.map_univ_coeEmbedding] at this
   simpa only [Finset.map_val, Multiset.coeEmbedding_apply, Multiset.map_map,
     Function.comp_apply] using this
 
 @[simp]
-theorem Multiset.map_univ {β : Type*} (m : Multiset α) (f : α → β) :
+theorem map_univ {β : Type*} (m : Multiset α) (f : α → β) :
     ((Finset.univ : Finset m).val.map fun (x : m) ↦ f (x : α)) = m.map f := by
   erw [← Multiset.map_map, Multiset.map_univ_coe]
 
 @[simp]
-theorem Multiset.card_toEnumFinset (m : Multiset α) : m.toEnumFinset.card = Multiset.card m := by
+theorem card_toEnumFinset (m : Multiset α) : m.toEnumFinset.card = Multiset.card m := by
   rw [Finset.card, ← Multiset.card_map Prod.fst m.toEnumFinset.val]
   congr
   exact m.map_toEnumFinset_fst
 
 @[simp]
-theorem Multiset.card_coe (m : Multiset α) : Fintype.card m = Multiset.card m := by
+theorem card_coe (m : Multiset α) : Fintype.card m = Multiset.card m := by
   rw [Fintype.card_congr m.coeEquiv]
   simp only [Fintype.card_coe, card_toEnumFinset]
 
 @[to_additive]
-theorem Multiset.prod_eq_prod_coe [CommMonoid α] (m : Multiset α) : m.prod = ∏ x : m, (x : α) := by
+theorem prod_eq_prod_coe [CommMonoid α] (m : Multiset α) : m.prod = ∏ x : m, (x : α) := by
   congr
   simp
 
 @[to_additive]
-theorem Multiset.prod_eq_prod_toEnumFinset [CommMonoid α] (m : Multiset α) :
+theorem prod_eq_prod_toEnumFinset [CommMonoid α] (m : Multiset α) :
     m.prod = ∏ x ∈ m.toEnumFinset, x.1 := by
   congr
   simp
 
 @[to_additive]
-theorem Multiset.prod_toEnumFinset {β : Type*} [CommMonoid β] (m : Multiset α) (f : α → ℕ → β) :
+theorem prod_toEnumFinset {β : Type*} [CommMonoid β] (m : Multiset α) (f : α → ℕ → β) :
     ∏ x ∈ m.toEnumFinset, f x.1 x.2 = ∏ x : m, f x x.2 := by
   rw [Fintype.prod_equiv m.coeEquiv (fun x ↦ f x x.2) fun x ↦ f x.1.1 x.1.2]
   · rw [← m.toEnumFinset.prod_coe_sort fun x ↦ f x.1 x.2]
   · intro x
     rfl
+
+/--
+If `s = t` then there's an equivalence between the appropriate types.
+-/
+@[simps]
+def cast {s t : Multiset α} (h : s = t) : s ≃ t where
+  toFun x := ⟨x.1, x.2.cast (by simp [h])⟩
+  invFun x := ⟨x.1, x.2.cast (by simp [h])⟩
+  left_inv x := rfl
+  right_inv x := rfl
+
+instance : IsEmpty (0 : Multiset α) := Fintype.card_eq_zero_iff.mp (by simp)
+
+instance : IsEmpty (∅ : Multiset α) := Fintype.card_eq_zero_iff.mp (by simp)
+
```

Comment:
I think these probably could live in an earlier file?

Reply:
This is the file where the coercion from multiset to `Type` is defined, and this depends on that